Add ShardCluster implementation

congqixia opened the pull request:
ShardCluster maintains shard replica meta information
It watches node & segment change events
Provides shard replica search/query services

Related #16359 
/kind enhancement

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

letian-jiang commented on internal/querynode/shard_cluster.go:
```diff
@@ -0,0 +1,477 @@
+	return segment, ok
+}
+
+// segmentAllocations returns segment to node mappings.
```
I guess it should be node to segments mapping.

congqixia replied:
@letian-jiang nice catch. Will fix in next pr.

congqixia commented on internal/querynode/shard_cluster.go:
```diff
@@ -0,0 +1,477 @@
+	// get node allocation
+	segAllocs := sc.segmentAllocations()
+
+	// TODO dispatch to local queryShardService query dml channel growing segments
```
@letian-jiang it depends on how we combine queryservice and shardCluster. Put this TODO as a reminder.

congqixia committed:
Add ShardCluster implementation

ShardCluster maintains shard replica meta information
It watches node & segment change events
Provides shard replica search/query services

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

letian-jiang commented:
/lgtm
